src/lseq.rs

```diff
@@ -0,0 +1,255 @@
+        dot: Dot<A>,
+    },
+}
+
+impl<T, A: Ident> LSeq<T, A> {
+    /// Create an LSEQ for the empty string
+    pub fn new(id: A) -> Self {
+        LSeq {
+            seq: Vec::new(),
+            gen: IdentGen::new(id.clone()),
+            dot: Dot::new(id, 0),
+        }
+    }
+
+    /// Insert an identifier and value in the LSEQ
+    pub fn insert(&mut self, ix: Identifier<A>, dot: Dot<A>, c: T) {
+        // Inserts only have an impact if the identifier is in the tree
+        if let Err(res) = self.seq.binary_search_by(|e| e.id.cmp(&ix)) {
+            self.seq.insert(res, Entry { id: ix, dot, c });
+        }
+    }
+
+    /// Remove an identifier from the LSEQ
+    pub fn delete(&mut self, ix: Identifier<A>) {
+        // Deletes only have an effect if the identifier is already in the tree
+        if let Ok(i) = self.seq.binary_search_by(|e| e.id.cmp(&ix)) {
+            self.seq.remove(i);
+        }
+    }
+
+    /// Perform a local insertion of an element at a given position.
+    /// If `ix` is greater than the length of the LSeq then it is appended to the end.
+    ///
+    /// # Panics
+    ///
+    /// * If the allocation of a new index was not between `ix` and `ix - 1`.
+    pub fn insert_index(&mut self, ix: usize, c: T) -> Op<T, A>
+    where
+        T: Clone,
+    {
+        let lower = self.gen.lower();
+        let upper = self.gen.upper();
+
+        // If we're inserting past the length of the LSEQ then it's the same as appending.
+        let ix_ident = if self.seq.len() <= ix {
+            let prev = self
+                .seq
+                .last()
+                .map(|Entry { id, .. }| id)
+                .unwrap_or_else(|| &lower);
+            self.gen.alloc(prev, &upper)
+        } else {
+            // To insert an element at position ix, we want it to appear in the sequence between
+            // ix - 1 and ix. To do this, retreive each bound defaulting to the lower and upper
+            // bounds respectively if they are not found.
+            let prev = match ix.checked_sub(1) {
+                Some(i) => &self.seq.get(i).unwrap().id,
+                None => &lower,
+            };
+            let next = self
+                .seq
+                .get(ix)
+                .map(|Entry { id, .. }| id)
+                .unwrap_or(&upper);
+            let a = self.gen.alloc(&prev, next);
+
+            assert!(prev < &a);
+            assert!(&a < next);
+            a
+        };
+        let op = Op::Insert {
+            id: ix_ident,
+            dot: self.dot.clone(),
+            c,
+        };
+        self.dot.counter += 1;
+        self.apply(op.clone());
+        op
+    }
+
+    /// Perform a local deletion at ix.
+    ///
+    /// If `ix` is out of bounds, i.e. `ix > self.len()`, then
+    /// the `Op` is not performed and `None` is returned.
+    pub fn delete_index(&mut self, ix: usize) -> Option<Op<T, A>>
+    where
+        T: Clone,
+    {
+        if ix >= self.seq.len() {
```

Comment:
This is actually a _very_ frequent edge case for text-editors, maybe there should be a `delete_or_last` that can be used